python/lsst/daf/butler/registry/versions.py

```diff
@@ -0,0 +1,321 @@
+        total digest). Digest checking is optional and can be disabled in
+        configuration if configured digest is an empty string, we should delay
+        activating that check until we have a stable implementation for this
+        method.
+        """
+
+        def tableSchemaRepr(table: sqlalchemy.schema.Table) -> str:
+            """Make string representation of a single table schema.
+            """
+            tableSchemaRepr = [table.name]
+            schemaReps = []
+            for column in table.columns:
+                columnRep = f"COL,{column.name},{column.type}"
+                if column.primary_key:
+                    columnRep += ",PK"
+                if column.nullable:
+                    columnRep += ",NULL"
+                schemaReps += [columnRep]
+            for fkConstr in table.foreign_key_constraints:
+                fkRep = f"FK,{fkConstr.name}"
+                for fk in fkConstr.elements:
+                    fkRep += f"{fk.column.name}->{fk.target_fullname}"
+                schemaReps += [fkRep]
+            schemaReps.sort()
+            tableSchemaRepr += schemaReps
+            return ";".join(tableSchemaRepr)
+
+        md5 = hashlib.md5()
+        tableSchemas = sorted(tableSchemaRepr(table) for table in tables)
+        for tableRepr in tableSchemas:
+            md5.update(tableRepr.encode())
+        digest = md5.hexdigest()
+        return digest
+
+    def addTable(self, group: str, table: sqlalchemy.schema.Table) -> None:
+        """Add a table to specified schema group.
+
+        Table schema added to a group will be used when calculating digest
+        for that group.
+
+        Parameters
+        ----------
+        group : `str`
+            Schema group name, e.g. "core", or " dimensions".
+        table : `sqlalchemy.schema.Table`
+            Table schema.
+        """
+        self._tablesGroups.setdefault(group, []). append(table)
+
+    def storeVersions(self, attributes: ButlerAttributeManager) -> None:
+        """Store configured schema versions in registry arttributes.
+
+        Parameters
+        ----------
+        attributes : `ButlerAttributeManager`
+            Attribute manager instance.
+        """
+        for key, vInfo in self._versions.items():
+            # attribute name reflects configuration path in "registry" config
+            attributes.set(f"schema_versions.{key}.version", str(vInfo.version))
+            # TODO: we could also store digest in the database but I'm not
+            # sure that digest calculation is stable enough at this point.
+
+    def checkVersionDigests(self) -> None:
+        """Compare current schema digest to a configured digest.
+
+        It calculates digest to all schema groups using tables added to each
+        group with `addTable` method. If digest is different from a configured
+        digest for the same group it generates logging warning message.
```

Comment:
_Only_ logging a warning seems like a fairly limited way to report problems.  Might make sense to have this at least return a `bool` as well so it could be used by callers that might want to raise exceptions instead.

Reply:
I'm not sure that simple `bool` would be useful for clients. The use case for generating warnings is that when you update you code and change schema you still do not know the digest for the new schema. To calculate new digest you would run the code and with give you a warning that is expected and a new digest that you can you to update your configuration. Returning False to client does not produce much useful info to decide what needs to be done, the code will likely run if your changes are OK, it's just the configuration does not match your code.

Reply:
And I could potentially add a separate application that calculates digests and then we could require that digest should always be correct in the configuration but that maybe too disruptive during development process (though it is possible to disable digest check now).

Reply:
If there's no real use case, that's okay.  I just found it a little weird that the only way an error was reported was through logs.

Reply:
I do not consider it an error yet, just a reminder for a developer that schema version may have changed. Digest calculation is not bullet-proof right now, if we can make it more reliable in the future we can proabably switch to exceptions or returning error status from this method.